File: packages/vc/src/proof.test.ts
import { readFileSync } from 'node:fs'
import { bytesToHex } from '@noble/hashes/utils.js'
import { describe, expect, it } from 'vitest'
import { canonicalize } from './canon.js'
import { staticLoader } from './context/loader.js'
import { decodeBase58btc, decodeMultikey } from './multibase.js'
import { createProof, hashForProof, memorySigner, verifyProof } from './proof.js'

const read = (name: string) => JSON.parse(readFileSync(new URL(`../../../vectors/w3c/${name}`, import.meta.url), 'utf8'))

// The test vector published in the W3C Data Integrity EdDSA Cryptosuites specification.
const w3c = read('eddsa-rdfc-2022.json')
const loader = staticLoader({ 'https://www.w3.org/ns/credentials/examples/v2': read('examples-v2.json') })

// secretKeyMultibase is a two-byte multicodec prefix followed by the 32-byte seed.
const seed = decodeBase58btc(w3c.secretKeyMultibase).slice(2)
const publicKey = decodeMultikey(w3c.publicKeyMultibase)

describe('W3C eddsa-rdfc-2022 known answers', () => {
  it('canonicalizes the document and the proof options as published', async () => {
    expect(await canonicalize(w3c.unsigned, loader)).toBe(w3c.canonicalDocument)
    expect(await canonicalize(w3c.proofOptions, loader)).toBe(w3c.canonicalProofOptions)
  })

  it('produces the published 64-byte hash', async () => {
    const { '@context': _, ...options } = w3c.proofOptions
    expect(bytesToHex(await hashForProof(w3c.unsigned, options, loader))).toBe(w3c.combinedHashHex)
  })

  it('derives the published public key from the secret key', async () => {
    const signer = await memorySigner(seed, w3c.proofOptions.verificationMethod)
    expect(signer.publicKey).toEqual(publicKey)
  })

  it('produces the published proofValue', async () => {
    const signer = await memorySigner(seed, w3c.proofOptions.verificationMethod)
    const signed = await createProof(w3c.unsigned, signer, { created: w3c.proofOptions.created }, loader)
    expect(signed.proof.proofValue).toBe(w3c.signed.proof.proofValue)
    expect(signed).toEqual(w3c.signed)
  })

  it('verifies the published credential', async () => {
    expect(await verifyProof(w3c.signed, publicKey, loader)).toBe(true)
  })
})

describe('verifyProof', () => {
  const tampered = (mutate: (c: any) => void) => {
    const copy = structuredClone(w3c.signed)
    mutate(copy)
    return copy
  }

  it.each([
    ['a changed subject value', (c: any) => (c.credentialSubject.alumniOf = 'The School of Exampless')],
    ['a changed issuer', (c: any) => (c.issuer = 'https://vc.example/issuers/5679')],
    ['a changed proof date', (c: any) => (c.proof.created = '2023-02-24T23:36:39Z')],
    ['a changed verification method', (c: any) => (c.proof.verificationMethod += 'x')],
    ['a removed member', (c: any) => delete c.description],
  ])('is false after %s', async (_, mutate) => {
    expect(await verifyProof(tampered(mutate), publicKey, loader)).toBe(false)
  })

  it('is false for another key', async () => {
    const other = await memorySigner(new Uint8Array(32).fill(7), 'did:example:other#key-1')
    expect(await verifyProof(w3c.signed, other.publicKey, loader)).toBe(false)
  })

  it.each([
    ['no proof', (c: any) => delete c.proof],
    ['a list of proofs', (c: any) => (c.proof = [c.proof])],
    ['a proofValue in another base', (c: any) => (c.proof.proofValue = 'u' + c.proof.proofValue.slice(1))],
    ['a truncated proofValue', (c: any) => (c.proof.proofValue = c.proof.proofValue.slice(0, 40))],
    ['a proofValue that is not text', (c: any) => (c.proof.proofValue = 42)],
  ])('is false for %s', async (_, mutate) => {
    expect(await verifyProof(tampered(mutate), publicKey, loader)).toBe(false)
  })
})

describe('createProof', () => {
  it('refuses a document that is already signed', async () => {
    const signer = await memorySigner(seed, 'did:example:x#key-1')
    await expect(createProof(w3c.signed, signer, { created: '2026-01-01T00:00:00Z' }, loader)).rejects.toThrow()
  })
})

File: vc-go/proof/proof_test.go
package proof

import (
	"context"
	"encoding/hex"
	"encoding/json"
	"os"
	"path/filepath"
	"reflect"
	"testing"
	"time"

	"github.com/piprate/json-gold/ld"

	"github.com/vemphy/vc/vc-go/canon"
	"github.com/vemphy/vc/vc-go/internal/vectors"
	"github.com/vemphy/vc/vc-go/multibase"
	"github.com/vemphy/vc/vc-go/vcctx"
)

// The test vector published in the W3C Data Integrity EdDSA Cryptosuites specification.
type w3cVector struct {
	PublicKeyMultibase    string         `json:"publicKeyMultibase"`
	SecretKeyMultibase    string         `json:"secretKeyMultibase"`
	Unsigned              map[string]any `json:"unsigned"`
	CanonicalDocument     string         `json:"canonicalDocument"`
	ProofOptions          map[string]any `json:"proofOptions"`
	CanonicalProofOptions string         `json:"canonicalProofOptions"`
	CombinedHashHex       string         `json:"combinedHashHex"`
	Signed                map[string]any `json:"signed"`
}

func load(t *testing.T) (w3cVector, ld.DocumentLoader) {
	t.Helper()
	var v w3cVector
	vectors.Load(t, "w3c/eddsa-rdfc-2022.json", &v)
	examples, err := os.ReadFile(filepath.Join(vectors.Dir(t), "w3c", "examples-v2.json"))
	if err != nil {
		t.Fatal(err)
	}
	return v, vcctx.Loader(map[string][]byte{"https://www.w3.org/ns/credentials/examples/v2": examples})
}

func signer(t *testing.T, v w3cVector) *MemorySigner {
	t.Helper()
	// secretKeyMultibase is a two-byte multicodec prefix followed by the 32-byte seed.
	secret, err := multibase.DecodeBase58btc(v.SecretKeyMultibase)
	if err != nil {
		t.Fatal(err)
	}
	s, err := NewMemorySigner(secret[2:], v.ProofOptions["verificationMethod"].(string))
	if err != nil {
		t.Fatal(err)
	}
	return s
}

func TestW3CCanonicalForms(t *testing.T) {
	v, loader := load(t)
	if got, err := canon.Canonicalize(v.Unsigned, loader); err != nil || got != v.CanonicalDocument {
		t.Errorf("document: %v\n%s", err, got)
	}
	if got, err := canon.Canonicalize(v.ProofOptions, loader); err != nil || got != v.CanonicalProofOptions {
		t.Errorf("proof options: %v\n%s", err, got)
	}
}

func TestW3CHash(t *testing.T) {
	v, loader := load(t)
	options := map[string]any{}
	for k, val := range v.ProofOptions {
		if k != "@context" {
			options[k] = val
		}
	}
	hash, err := Hash(v.Unsigned, options, loader)
	if err != nil {
		t.Fatal(err)
	}
	if got := hex.EncodeToString(hash); got != v.CombinedHashHex {
		t.Errorf("hash %s, want %s", got, v.CombinedHashHex)
	}
}

func TestW3CPublicKey(t *testing.T) {
	v, _ := load(t)
	want, _ := multibase.DecodeMultikey(v.PublicKeyMultibase)
	if got := signer(t, v).PublicKey(); !got.Equal(want) {
		t.Errorf("public key %x, want %x", got, want)
	}
}

func TestW3CProofValue(t *testing.T) {
	v, loader := load(t)
	created, _ := time.Parse(time.RFC3339, v.ProofOptions["created"].(string))
	signed, err := Create(context.Background(), v.Unsigned, signer(t, v), created, loader)
	if err != nil {
		t.Fatal(err)
	}
	if !reflect.DeepEqual(signed, v.Signed) {
		got, _ := json.MarshalIndent(signed["proof"], "", "  ")
		t.Errorf("signed credential differs from the published one; proof:\n%s", got)
	}
}

func TestW3CVerify(t *testing.T) {
	v, loader := load(t)
	key, _ := multibase.DecodeMultikey(v.PublicKeyMultibase)
	if ok, err := Verify(v.Signed, key, loader); err != nil || !ok {
		t.Errorf("Verify = %v, %v", ok, err)
	}
}

func clone(t *testing.T, doc map[string]any) map[string]any {
	t.Helper()
	raw, _ := json.Marshal(doc)
	var out map[string]any
	if err := json.Unmarshal(raw, &out); err != nil {
		t.Fatal(err)
	}
	return out
}

func TestVerifyIsFalseAfterAnyChange(t *testing.T) {
	v, loader := load(t)
	key, _ := multibase.DecodeMultikey(v.PublicKeyMultibase)
	proofOf := func(c map[string]any) map[string]any { return c["proof"].(map[string]any) }

	cases := map[string]func(c map[string]any){
		"changed subject value": func(c map[string]any) {
			c["credentialSubject"].(map[string]any)["alumniOf"] = "The School of Exampless"
		},
		"changed issuer":              func(c map[string]any) { c["issuer"] = "https://vc.example/issuers/5679" },
		"changed proof date":          func(c map[string]any) { proofOf(c)["created"] = "2023-02-24T23:36:39Z" },
		"changed verification method": func(c map[string]any) { proofOf(c)["verificationMethod"] = "did:example:other#key-1" },
		"removed member":              func(c map[string]any) { delete(c, "description") },
		"no proof":                    func(c map[string]any) { delete(c, "proof") },
		"list of proofs":              func(c map[string]any) { c["proof"] = []any{c["proof"]} },
		"proofValue in another base":  func(c map[string]any) { proofOf(c)["proofValue"] = "u" + proofOf(c)["proofValue"].(string)[1:] },
		"truncated proofValue":        func(c map[string]any) { proofOf(c)["proofValue"] = proofOf(c)["proofValue"].(string)[:40] },
		"proofValue that is not text": func(c map[string]any) { proofOf(c)["proofValue"] = 42 },
	}
	for name, mutate := range cases {
		c := clone(t, v.Signed)
		mutate(c)
		if ok, err := Verify(c, key, loader); ok || err != nil {
			t.Errorf("%s: Verify = %v, %v; want false, nil", name, ok, err)
		}
	}

	other, _ := NewMemorySigner(make([]byte, 32), "did:example:other#key-1")
	if ok, _ := Verify(v.Signed, other.PublicKey(), loader); ok {
		t.Error("verified under another key")
	}
}

func TestCreateRefusesSignedDocument(t *testing.T) {
	v, loader := load(t)
	if _, err := Create(context.Background(), v.Signed, signer(t, v), time.Now(), loader); err == nil {
		t.Error("signed a document that already has a proof")
	}
}
